File: audio-javafx/src/main/java/dev/kkorolyov/pancake/audio/javafx/JavaFXAudioFactory.java
package dev.kkorolyov.pancake.audio.javafx;

import dev.kkorolyov.pancake.platform.media.audio.Audio;
import dev.kkorolyov.pancake.platform.media.audio.AudioFactory;

import javafx.scene.media.Media;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * {@link AudioFactory} implemented through JavaFX.
 */
public class JavaFXAudioFactory implements AudioFactory {
	private final Map<String, Media> mediaCache = new ConcurrentHashMap<>();

	@Override
	public Audio get(String uri) {
		return new JavaFXAudio(mediaCache.computeIfAbsent(uri, Media::new));
	}
}

File: javafx-audio/src/main/kotlin/dev/kkorolyov/pancake/javafx/audio/JavaFxAudioFactory.kt
package dev.kkorolyov.pancake.javafx.audio

import dev.kkorolyov.pancake.platform.media.audio.Audio
import dev.kkorolyov.pancake.platform.media.audio.AudioFactory
import dev.kkorolyov.simplefuncs.function.Memoizer.memoize
import javafx.scene.media.Media

/**
 * [AudioFactory] implemented through JavaFX.
 */
class JavaFxAudioFactory : AudioFactory {
	private val mediaCache: (String) -> Media = memoize<String, Media>(::Media)::apply

	override fun get(uri: String): Audio = JavaFxAudio(mediaCache(uri))
}
